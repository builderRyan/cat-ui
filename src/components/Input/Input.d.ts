export type Ttheme = 'line' | 'line-pwd' | 'box' | 'box-pwd' | 'textarea';
export type TSize = 'large' | 'normal' | 'small';
export type TinputState = number;
export type THandleDOM = HTMLInputElement | HTMLTextAreaElement

export interface HandleProps {
  handleFocus?: (e: React.FocusEvent<THandleDOM>) => void;
  handleBlur?: (e: React.FocusEvent<THandleDOM>) => void;
  handleChange?: (e: React.ChangeEvent<THandleDOM>) => void;
  handleKeyDown?: (e: React.KeyboardEvent<THandleDOM>) => void;
  handleMouseEnter?: (e: React.MouseEvent<THandleDOM>) => void;
  handleMouseLeave?: (e: React.MouseEvent<THandleDOM>) => void;
  handleClipboard?: (e: React.MouseEvent<THandleDOM, MouseEvent> | React.ClipboardEvent<THandleDOM>) => void;
}

declare const enum inputStates {
  default,
  hover,
  active,
  error
}
