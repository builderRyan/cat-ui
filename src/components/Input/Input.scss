@import '../../style/index.scss';

.#{$prefix}-input {

}