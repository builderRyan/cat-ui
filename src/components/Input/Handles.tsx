import * as React from 'react';
import { InputProps, InputStates, inputStates } from './Input';
import TextArea, { TextAreaProps, TextAreaStates } from './TextArea';
import { THandleDOM } from './Input.d';

type HandlesProps = InputProps | TextAreaProps;
type HandlesStates = InputStates | TextAreaStates;

const Handles = (WrappedComponent: React.ComponentType<HandlesProps>) => {
  return class Input extends React.PureComponent<HandlesProps, HandlesStates> {
    static TextArea: typeof TextArea;

    public handleFocus(e: React.FocusEvent<THandleDOM>): void {
      const { inputState } = this.state;
      // set input active state
      inputState !== inputStates.error &&
        this.setState({
          inputState: inputStates.active
        });

      const { onFocus } = this.props;
      onFocus && onFocus(e as any);
    }

    public handleBlur(e: React.FocusEvent<THandleDOM>): void {
      const { inputState } = this.state;
      // set input default state
      inputState !== inputStates.error &&
        this.setState({
          inputState: inputStates.default
        });

      const { onBlur } = this.props;
      onBlur && onBlur(e as any);
    }

    public handleChange(e: React.ChangeEvent<THandleDOM>): void {
      const value = e.target.value;
      if (!('value' in this.props)) {
        this.setState({ value });
      }

      const { onChange } = this.props;
      onChange && onChange(e as any);
    }

    public handleKeyDown(e: React.KeyboardEvent<THandleDOM>): void {
      const { onPressEnter, onKeyDown } = this.props;
      if (e.keyCode === 13 && onPressEnter) {
        onPressEnter(e as any);
      }
      if (onKeyDown) {
        onKeyDown(e as any);
      }
    }

    public handleMouseEnter(e: React.MouseEvent<THandleDOM>): void {
      const { inputState } = this.state;
      inputState === inputStates.default &&
        this.setState({
          inputState: inputStates.hover
        });
    }

    public handleMouseLeave(e: React.MouseEvent<THandleDOM>): void {
      const { inputState } = this.state;
      inputState === inputStates.hover &&
        this.setState({
          inputState: inputStates.default
        });
    }

    public handleClipboard(
      e:
        | React.MouseEvent<THandleDOM, MouseEvent>
        | React.ClipboardEvent<THandleDOM>
    ): void {
      const {
        onPaste,
        onContextMenu,
        onCopy,
        onCut,
        clipboardFree,
        copyFree,
        pasteFree,
        cutFree
      } = this.props;

      switch (('' + e.type).toLowerCase()) {
        case 'copy':
          onCopy && onCopy(e as any);
          !copyFree && notFree();
          break;
        case 'paste':
          onPaste && onPaste(e as any);
          !pasteFree && notFree();
          break;
        case 'cut':
          onCut && onCut(e as any);
          !cutFree && notFree();
          break;
        case 'contextmenu':
          onContextMenu && onContextMenu(e as any);
          break;
      }

      !clipboardFree && notFree();

      function notFree() {
        if (e && e.preventDefault) {
          e.preventDefault();
        } else {
          (window.event as any).returnValue = false;
        }
      }
    }

    render() {
      return (
        <WrappedComponent
          {...this.props}
          handleFocus={this.handleFocus}
          handleBlur={this.handleBlur}
          handleChange={this.handleChange}
          handleKeyDown={this.handleKeyDown}
          handleMouseEnter={this.handleMouseEnter}
          handleMouseLeave={this.handleMouseLeave}
          handleClipboard={this.handleClipboard}
        />
      );
    }
  };
};

export default Handles;
