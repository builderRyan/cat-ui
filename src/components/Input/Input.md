Example

```jsx
import Input from './Input'

class Example extends React.Component {
  constructor (props) {
    super(props)
    this.state = {}
  }

  render () {
    return (<div className='input-box'>
      <span className='hint' >
        线型输入框 small:
      </span>
      <Input size='small' />
      <span className='hint' >
        线型输入框 small(不能复制):
      </span>
      <Input size='small' copyFree={false} />
      <span className='hint' >
        线型输入框 small(不能粘贴):
      </span>
      <Input size='small' pasteFree={false} />
      <span className='hint' >
        线型输入框 small(不能剪贴):
      </span>
      <Input size='small' cutFree={false} />
      <span className='hint' >
        线型输入框 small(不能复制+粘贴+剪贴):
      </span>
      <Input size='small' clipboardFree={false} />
      <span className='hint' >
        线型输入框 normal(带自动聚焦 未启用):
      </span>
      <Input size='normal' autoFocus={false} />
      <span className='hint' >
        线型输入框 large:
      </span>
      <Input size='large' />
      <span className='hint'>
        线型输入框禁用:
      </span>
      <Input className='item' disabled />
      <span className='hint' >
        线型输入框带清除按钮:
      </span>
      <Input className='item' showClear />
      <span className='hint' >
        线型密码输入框:
      </span>
      <Input className='item' type='line-pwd' />
      <span className='hint' >
        线型密码输入框带眼睛按钮:
      </span>
      <Input className='item' type='line-pwd' showEye />
      <span className='hint' >
        线型输入框带占位符:
      </span>
      <Input className='item' placeholder='Please fill something' />
      <span className='hint' >
        线型输入框带原生占位符(无动效):
      </span>
      <Input className='item' placeholder='Please fill something' placeholderOrigin />
      <span className='hint' >
        线型输入框带默认值:
      </span>
      <Input className='item' defaultValue='defaultValue' />
      <span className='hint' >
        线型输入框错误展示:
      </span>
      <Input className='item err' value='test value' error message='error msg is xxxx' />
      <span className='hint' >
        盒子输入框 small:
      </span>
      <Input size='small' type='box' />
      <span className='hint' >
        盒子输入框 normal:
      </span>
      <Input size='normal' type='box' />
      <span className='hint' >
        盒子输入框 large:
      </span>
      <Input size='large' type='box' />
      <span className='hint' >
        盒子输入框带清除按钮:
      </span>
      <Input className='item item-box' type='box' showClear />
      <span className='hint' >
        盒子密码输入框:
      </span>
      <Input className='item item-box' type='box-pwd' />
      <span className='hint' >
        盒子密码输入框带眼睛按钮:
      </span>
      <Input className='item item-box' type='box-pwd' showEye />
      <span className='hint' >
        盒子输入框带占位符:
      </span>
      <Input className='item item-box' type='box' placeholder='Please fill number' />
      <span className='hint' >
        盒子输入框带原生占位符(无动效):
      </span>
      <Input className='item item-box' type='box' placeholder='Please fill number' placeholderOrigin />
      <span className='hint' >
        盒子输入框带默认值:
      </span>
      <Input className='item item-box' type='box' defaultValue='you are awesome' />
      <span className='hint' >
        盒子输入框带错误展示:
      </span>
      <Input className='item item-box err' type='box' value='cuole' error message='error msg is zzzz' />
      <span className='hint' >
        盒子输入框带错误展示和占位符:
      </span>
      <Input className='item item-box err' type='box' value='123123' error message='error msg is yyyy' placeholder='Please fill username' />
      <span className='hint' >
        文本域:
      </span>
      <Input.TextArea className='item' />
      <style>
        {`
          .input-box {
            padding: 5px;
          }
          .input-box .item{
            display: block;
            width: 200px;
            height: 50px;
          }
          .input-box .hint{
            display: block;
            margin: 15px 0;
          }
          .input-box .item + .item{
            margin: 20px 0 0 0;
          }
          .input-box .item.err{
            margin-bottom: 50px;
          }
          .input-box .item.item-box{
            width: 380px;
            height: 52px;
          }
        `}
      </style>
    </div>
    )
  }
}

<Example />
```
