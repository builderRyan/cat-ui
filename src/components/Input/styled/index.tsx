import styled, { keyframes } from 'styled-components'
import { colorQuaternary, colorTertiary, colorPrimary, colorRedPrimary } from '../../../styles/theme'
import { IdomProps, inputStates } from '../Input'
import { IWrapperProps } from '../Wrapper'

const size_width = {
  'large': '400px',
  'normal': '300px',
  'small': '200px'
}

const size_height = {
  'large': '50px',
  'normal': '40px',
  'small': '30px'
}

const StyledInput = styled.input<IdomProps>`
  display: block;
  box-sizing: border-box;
  width: 100%;
  height: 100%;
  font-size: 16px;
  font-weight: normal;
  position: relative;
  z-index: 1;
  background-color: transparent;
  color: ${colorPrimary};
  border: 0;
  outline: 0;
  &:focus{
    outline: none;
  }
  &::-ms-clear{
    display:none;
  }
  &::-ms-reveal{
    display:none;
  }
  &::-webkit-credentials-auto-fill-button {
    visibility: hidden;
    display: none !important;
    pointer-events: none;
    position: absolute;
    right: 0;
  }
  &::placeholder,
  &::-webkit-input-placeholder{
    font-size: 16px;
    font-weight: normal;
    line-height: 1.5;
    color: ${colorTertiary};
  };
  cursor: ${props => props.disabled && 'not-allowed'}
`

const StyledIcon = styled.span<IWrapperProps>`
  cursor: pointer;
  position: absolute;
  top: 50%
  transform: translate(0, -50%);
  z-index: 1;
  width: 10px;
  height: 10px;
  right: 5.6%;
  color: ${props => (props.state === inputStates.error ? colorRedPrimary : colorPrimary)};
  font-size: ${props => props.showClear ? '10px' : '16px'};
`

const StyledWrapper = styled.div<IWrapperProps>`
  display: inline-block;
  width: ${props => props.size ? (size_width[props.size] || '100%') : '100%'};
  height: ${props => props.size ? (size_height[props.size] || '100%') : '100%'};
  position: relative;
  box-sizing: border-box;
  padding-left: ${props => (props.theme === 'box' || props.theme === 'box-pwd' ? '20px' : 0)};
  padding-right: ${props => props.showClear || props.showEye ? '3.5%' : 0};
`

const StyledWrapperBorder = styled.span<IWrapperProps>`
  box-sizing: border-box;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  transition: all .3s;
  border-left: ${props => {
    if (props.theme === 'line' || props.theme === 'line-pwd') return 0
    return '1px solid'
  }};
  border-right: ${props => {
    if (props.theme === 'line' || props.theme === 'line-pwd') return 0
    return '1px solid'
  }};
  border-top: ${props => {
    if (props.theme === 'line' || props.theme === 'line-pwd') return 0
    return '1px solid'
  }};
  border-bottom: 1px solid;
  border-color: ${props => {
    let color
    switch (props.state) {
      case inputStates.default:
        color = colorQuaternary
        break
      case inputStates.hover:
        color = colorTertiary
        break
      case inputStates.active:
        color = colorPrimary
        break
      case inputStates.error:
        color = colorRedPrimary
        break
      default:
        color = colorQuaternary
    };
    return color
  }};
  border-radius: ${props => props.theme === 'textarea' ? '2px' : (props.theme === 'box' || props.theme === 'box-pwd' ? '4px' : 0)};
`

const flyout = keyframes`
  from {
    font-size: 16px;
    color: ${colorTertiary};
    background-color: transparent;
    transform: translate(0, -50%);
    top: 50%;
  }

  to {
    font-size: 12px;
    color: ${colorPrimary};
    background-color: #ffffff;
    transform: translate(0, 0);
    top: -15%;
  }
`;

const flyin = keyframes`
  from {
    font-size: 12px;
    color: ${colorPrimary};
    background-color: #ffffff;
    transform: translate(0, 0);
    top: -15%;
  }

  to {
    font-size: 16px;
    color: ${colorTertiary};
    background-color: transparent;
    transform: translate(0, -50%);
    top: 50%;
  }
`;

const StyledWrapperPlaceholder = styled.span<IWrapperProps>`
  position: absolute;
  top: 50%
  z-index: 1;
  left: ${props => (props.theme === 'box' || props.theme === 'box-pwd' ? '15px' : 0)};
  padding: ${props => (props.theme === 'box' || props.theme === 'box-pwd' ? '0 5px' : 0)};
  transform: translate(0, -50%);
  font-size: 16px;
  font-weight: normal;
  line-height: 1;
  letter-spacing: 0.75px;
  color: ${colorTertiary};
  transition: all .3s;
  animation: ${props => ((props.state === inputStates.active || props.value) ? flyout : flyin)} .2s ease-out forwards;
`

const StyledWrapperMsg = styled.span<IWrapperProps>`
  display: block;
  font-size: 12px;
  color: ${props => (props.state === inputStates.error ? colorRedPrimary : colorPrimary)};
  font-weight: normal;
  margin-top: 8px;
`

const StyledTextArea = styled.textarea<IWrapperProps>`
  display: block;
  box-sizing: border-box;
  width: 100%;
  height: 100%;
  font-size: 16px;
  font-weight: normal;
  position: relative;
  z-index: 1;
  background-color: transparent;
  color: ${colorPrimary};
  border: 0;
  outline: 0;
  &:focus{
    outline: none;
  }
`

export { StyledInput, StyledIcon, StyledWrapper, StyledWrapperBorder, StyledWrapperPlaceholder, StyledWrapperMsg, StyledTextArea }