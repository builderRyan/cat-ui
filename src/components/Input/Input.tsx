import * as React from 'react'
import { pick } from 'lodash'
import { StyledInput, StyledIcon } from './styled'
import { TSize, Ttheme, TinputState, HandleProps } from './Input.d'
import Wrapper from './Wrapper'
import Handles from './Handles'
import Icon from '../Icon'

export interface InputProps extends HandleProps {
  /** 自动聚焦 */
  autoFocus?: boolean;
  /** 类名 */
  className?: string;
  /** 默认值 */
  defaultValue?: string;
  /** 禁用 */
  disabled?: boolean;
  /** 值 - 如果传入value属性，则input框变为受控, 因此输入框的值只根据value设置 */
  value?: string;
  /** 大小尺寸 'large' | 'normal' | 'small' */
  size?: TSize;
  /** 类型 'line' | 'line-pwd' | 'box' | 'box-pwd' */
  type?: Ttheme;
  /** 占位符 */
  placeholder?: string;
  /** 是否启用原生占位符(无动效) */
  placeholderOrigin?: boolean;
  /** 展示一键清除按钮 */
  showClear?: boolean;
  /** 展示密码和文本切换按钮 */
  showEye?: boolean;
  /** 是否出错 */
  error?: boolean;
  /** 输入框消息提示 */
  message?: string;
  /** 是否能对输入框进行复制、粘贴、剪贴的操作 */
  clipboardFree?: boolean;
  /** 是否能对输入框进行复制的操作 */
  copyFree? : boolean;
  /** 是否能对输入框进行粘贴的操作 */
  pasteFree? : boolean;
  /** 是否能对输入框进行剪贴的操作 */
  cutFree? : boolean;
  /** icon style样式 */
  iconStyle?: object;
  /** 聚焦回调 */
  onFocus?: (e: React.FocusEvent<HTMLInputElement>) => any;
  /** 失焦回调 */
  onBlur?: (e: React.FocusEvent<HTMLInputElement>) => any;
  /** 值改变的回调 */
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => any;
  /** 按键的回调 */
  onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement>) => any;
  /** 按下回车键的回调 */
  onPressEnter?: (e: React.KeyboardEvent<HTMLInputElement>) => any;
  /** 执行粘贴操作的回调 */
  onPaste?: (e: React.MouseEvent<HTMLInputElement, MouseEvent> | React.ClipboardEvent<HTMLInputElement>) => any;
  /** 点击鼠标右键的回调 */
  onContextMenu?: (e: React.MouseEvent<HTMLInputElement, MouseEvent> | React.ClipboardEvent<HTMLInputElement>) => any;
  /** 执行复制操作的回调 */
  onCopy?: (e: React.MouseEvent<HTMLInputElement, MouseEvent> | React.ClipboardEvent<HTMLInputElement>) => any;
  /** 执行剪切操作的回调 */
  onCut?: (e: React.MouseEvent<HTMLInputElement, MouseEvent> | React.ClipboardEvent<HTMLInputElement>) => any;
}

export interface InputStates {
  type: 'text' | 'password';
  value: string;
  domProps: Array<string>;
  inputState: TinputState;
}

export interface IdomProps {
  disabled?: boolean;
  [propName: string]: any;
}

const domProps: IdomProps = {
  disabled: false
}

export const enum inputStates {
  default,
  hover,
  active,
  error
}

class Input extends React.PureComponent<InputProps, InputStates> {
  static defaultProps = {
    autoFocus: false,
    disabled: false,
    type: 'line',
    error: false,
    placeholderOrigin: false,
    clipboardFree: true,
    copyFree: true,
    pasteFree: true,
    cutFree: true
  }

  static getDerivedStateFromProps(nextProps: InputProps) {
    if ('value' in nextProps) {
      return {
        value: nextProps.value
      }
    }
    return null
  }

  private input: React.RefObject<HTMLInputElement>;
  private isIconBlur: boolean;

  public constructor (props: InputProps) {
    super(props)
    const type = typeof props.type === 'undefined' ? 'text' : (props.type === 'line-pwd' || props.type === 'box-pwd' ? 'password' : 'text')
    const value = (typeof props.value === 'undefined' ? props.defaultValue : props.value) || ''
    const inputState = props.error ? inputStates.error : inputStates.default
    this.state = {
      type,
      value,
      domProps: Object.keys(domProps),
      inputState
    }

    this.input = React.createRef()
    this.isIconBlur = false
  }

  public componentDidMount () {
    const { autoFocus } = this.props
    if (autoFocus) this.input.current.focus()
  }

  private handleClear (e: React.MouseEvent<HTMLElement, MouseEvent>): void {
    this.setState({ value: '' })
    const { onChange } = this.props
    if (onChange) {
      let event = e
      event = Object.create(e)
      event.target = this.input.current
      event.currentTarget = this.input.current
      const originalInputValue = this.input.current.value
      this.input.current.value = ''
      onChange(event as any)
      this.input.current.value = originalInputValue
    }
  }

  private handleEye (): void {
    this.setState(prevState => {
      const type = prevState.type === 'text' ? 'password' : 'text'
      return { type }
    })
  }

  private handleBlur (e: React.FocusEvent<HTMLInputElement>): void {
    if (!this.isIconBlur) {
      const { handleBlur } = this.props
      handleBlur.call(this, e)
    }
  }

  private iconForbidBlur (): void {
    this.isIconBlur = true
  }

  private iconAllowBlur (): void {
    this.input.current.focus()
    this.isIconBlur = false
  }

  public render () {
    const { type, value, domProps, inputState } = this.state
    const { size, className, message, placeholder, placeholderOrigin, showClear, showEye, iconStyle,
      handleFocus, handleChange, handleKeyDown, handleMouseEnter, handleMouseLeave, handleClipboard
    } = this.props
    const theme = this.props.type
    const props = pick(this.props, domProps) as IdomProps

    return (
      <Wrapper
        value={value}
        theme={theme}
        size={size}
        state={inputState}
        className={className}
        placeholder={!placeholderOrigin && placeholder}
        message={message}
        showClear={showClear}
        showEye={showEye}
      >
        <StyledInput
          {...props}
          type={type}
          value={value}
          onFocus={handleFocus.bind(this)}
          onBlur={this.handleBlur.bind(this)}
          onChange={handleChange.bind(this)}
          onKeyDown={handleKeyDown.bind(this)}
          onMouseEnter={handleMouseEnter.bind(this)}
          onMouseLeave={handleMouseLeave.bind(this)}
          onPaste={handleClipboard.bind(this)}
          onContextMenu={handleClipboard.bind(this)}
          onCopy={handleClipboard.bind(this)}
          onCut={handleClipboard.bind(this)}
          placeholder={placeholderOrigin ? placeholder : ''}
          ref={this.input}
        />
        <StyledIcon
          showClear={showClear}
          showEye={showEye}
          style={iconStyle}
          onMouseDown={this.iconForbidBlur.bind(this)}
          onMouseUp={this.iconAllowBlur.bind(this)}
          onTouchStart={this.iconForbidBlur.bind(this)}
          onTouchEnd={this.iconAllowBlur.bind(this)}
        >
          {showClear && <Icon type='close' onClick={this.handleClear.bind(this)} />}
          {!showClear && (theme === 'line-pwd' || theme === 'box-pwd') && showEye && <Icon type={type === 'text' ? 'eye-open' : 'eye-close'} onClick={this.handleEye.bind(this)} />}
        </StyledIcon>
      </Wrapper>
    )
  }
}

export default Handles(Input)
