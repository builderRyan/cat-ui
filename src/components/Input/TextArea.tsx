import * as React from 'react'
import { pick } from 'lodash'
import { IdomProps, inputStates } from './Input'
import { TinputState, HandleProps } from './Input.d'
import { StyledTextArea } from './styled'
import Wrapper from './Wrapper'
import Handles from './Handles'

export interface TextAreaProps extends HandleProps {
  /** 自动聚焦 */
  autoFocus?: boolean;
  /** 类名 */
  className?: string;
  /** 默认值 */
  defaultValue?: string;
  /** 禁用 */
  disabled?: boolean;
  /** 值 - 如果传入value属性，则input框变为受控, 因此输入框的值只根据value设置 */
  value?: string;
  /** 是否出错 */
  error?: boolean;
  /** 输入框消息提示 */
  message?: string;
  /** 自动大小计算 */
  autosize?: boolean;
  /** 显示计数 */
  showCount?: boolean;
  /** 最大长度 */
  maxlength?: number;
  /** 是否只读 */
  readonly?: boolean;
  /** 是否能对输入框进行复制、粘贴、剪贴的操作 */
  clipboardFree?: boolean;
  /** 是否能对输入框进行复制的操作 */
  copyFree? : boolean;
  /** 是否能对输入框进行粘贴的操作 */
  pasteFree? : boolean;
  /** 是否能对输入框进行剪贴的操作 */
  cutFree? : boolean;
  /** 聚焦回调 */
  onFocus?: (e: React.FocusEvent<HTMLTextAreaElement>) => any;
  /** 失焦回调 */
  onBlur?: (e: React.FocusEvent<HTMLTextAreaElement>) => any;
  /** 值改变的回调 */
  onChange?: (e: React.ChangeEvent<HTMLTextAreaElement>) => any;
  /** 按键的回调 */
  onKeyDown?: (e: React.KeyboardEvent<HTMLTextAreaElement>) => any;
  /** 按下回车键的回调 */
  onPressEnter?: (e: React.KeyboardEvent<HTMLTextAreaElement>) => any;
  /** 执行粘贴操作的回调 */
  onPaste?: (e: React.MouseEvent<HTMLTextAreaElement, MouseEvent> | React.ClipboardEvent<HTMLTextAreaElement>) => any;
  /** 点击鼠标右键的回调 */
  onContextMenu?: (e: React.MouseEvent<HTMLTextAreaElement, MouseEvent> | React.ClipboardEvent<HTMLTextAreaElement>) => any;
  /** 执行复制操作的回调 */
  onCopy?: (e: React.MouseEvent<HTMLTextAreaElement, MouseEvent> | React.ClipboardEvent<HTMLTextAreaElement>) => any;
  /** 执行剪切操作的回调 */
  onCut?: (e: React.MouseEvent<HTMLTextAreaElement, MouseEvent> | React.ClipboardEvent<HTMLTextAreaElement>) => any;
}

export interface TextAreaStates {
  value: string;
  inputState: TinputState;
}

const domProps: IdomProps = {
  disabled: false,
  maxlength: 500,
  placeholder: '',
  readonly: false
}

class TextArea extends React.PureComponent<TextAreaProps, TextAreaStates> {
  static defaultProps = {
    autosize: false,
    disabled: false,
    error: false,
    placeholderOrigin: false,
    clipboardFree: true,
    copyFree: true,
    pasteFree: true,
    cutFree: true
  }

  constructor (props: TextAreaProps) {
    super(props)
    const value = (typeof props.value === 'undefined' ? props.defaultValue : props.value) || ''
    const inputState = props.error ? inputStates.error : inputStates.default
    this.state = {
      value,
      inputState
    }
  }

  render () {
    const { value, inputState } = this.state
    const { className, autosize, showCount, maxlength, handleFocus, handleBlur, handleChange, handleKeyDown, handleMouseEnter, handleMouseLeave, handleClipboard } = this.props
    const props = pick(this.props, domProps) as IdomProps

    return (
      <Wrapper
        value={value}
        theme='textarea'
        state={inputState}
        className={className}
        showCount={showCount}
        maxlength={maxlength}
      >
        <StyledTextArea
          {...props}
          onFocus={handleFocus.bind(this)}
          onBlur={handleBlur.bind(this)}
          onChange={handleChange.bind(this)}
          onKeyDown={handleKeyDown.bind(this)}
          onMouseEnter={handleMouseEnter.bind(this)}
          onMouseLeave={handleMouseLeave.bind(this)}
          onPaste={handleClipboard.bind(this)}
          onContextMenu={handleClipboard.bind(this)}
          onCopy={handleClipboard.bind(this)}
          onCut={handleClipboard.bind(this)}
        />
      </Wrapper>
    )
  }
}

export default Handles(TextArea)
