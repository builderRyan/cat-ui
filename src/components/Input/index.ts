import Input from './Input'
import TextArea from './TextArea'

Input.TextArea = TextArea

// export { default as TextArea } from './TextArea'
export default Input
