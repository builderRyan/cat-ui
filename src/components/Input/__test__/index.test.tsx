import * as React from 'react'
import { render, mount } from 'enzyme'
import Input from '../index'

describe('Input', () => {
  test('renders correctly', () => {
    const wrapper = render(
      <Input />
    )
    expect(wrapper).toMatchSnapshot()
  })

})
