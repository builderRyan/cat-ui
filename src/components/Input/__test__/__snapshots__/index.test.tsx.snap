// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "hmly-input-label hmly-input-line-label sc-bdVaJa bSrzPW",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "hmly-input hmly-input-line sc-bwzfXH lgcGxt",
        },
        "children": Array [],
        "name": "input",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "label",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
